### ISSUE TITLE
NewMatFromBytes directly uses an unsafe pointer to the []byte, but no reference preventing garbage collection

### ISSUE THREAD

**Issue by dougnd** [hybridgroup/gocv#285]
## Description

The current NewMatFromBytes (in gocv 0.15) does not copy the data from the slice but instead uses an unsafe pointer to the data when initializing the OpenCV Mat.  Furthermore, the gocv Mat maintains no reference or anything to this data to prevent that slice from being garbage collected (the relationship between the Mat and the byte slice may not even be picked up by the escape analysis -- perhaps the byte slices are being allocated on the stack!).

I don't know if this was expected behavior or not, but I was caught off guard!  If this is working as designed, kindly leave a few notes in the documentation for NewMatFromBytes.  

My workaround for now is to clone the Mat after calling NewMatFromBytes (before the []byte goes out of scope).

## Steps to Reproduce

Here is a test case that reproduces the strange problems I was experiencing (Mat's getting overridden after the byte slice is out of scope).

```
import (
	"crypto/rand"
	"testing"

	"github.com/stretchr/testify/require"
	"gocv.io/x/gocv"
)

// Equal returns whether the images are equal.
func Equal(img1, img2 gocv.Mat) bool {
	eqMat := gocv.NewMat()
	gocv.Compare(img1, img2, &eqMat, gocv.CompareNE)
	count := gocv.CountNonZero(eqMat)
	eqMat.Close()
	return count == 0
}

func TestMatFromBytes(t *testing.T) {
	size := 500
	var img, imgCloned gocv.Mat
	var err error

	makeImg := func() {
		b := make([]byte, size*size)
		_, err = rand.Read(b)
		img, err = gocv.NewMatFromBytes(size, size, gocv.MatTypeCV8U, b)
		require.NoError(t, err)
		imgCloned = img.Clone()
	}

	useMemory := func() {
		b := make([]byte, size*size)
		_, err = rand.Read(b)
	}

	makeImg()

	// This assertion will pass.
	require.True(t, Equal(img, imgCloned))
	for i := 0; i < 100; i++ {
		useMemory()
	}

	// This assertion will fail!
	require.True(t, Equal(img, imgCloned))
}

```

## Your Environment

* Operating System and version: Mac OSX 10.13.6
* OpenCV version used: 3.4.2
* How did you install OpenCV? (homebrew)
* GoCV version used: 0.15.0
* Go version: 1.10.3
* Did you run the `env.sh` or `env.cmd` script before trying to `go run` or `go build`? (N/A?)

P.S.  OpenCV in go is awesome and I look forward to this project becoming mature!  Thanks for all the work you guys do!

**Comment by deadprogram** [comment]
Hi @dougnd sorry it is taking a bit to get back to you on this issue. You bring up some interesting points, I need to consider the implications briefly. Anyone else want to comment here?

**Comment by netbrain** [comment]
Any progress on this? I believe i just hit an issue that is directly related to this. 

In my application I have a cache of preprocessed mats that are re-used under certain conditions. However after a while the video feed that is processed gets distorted and i believe the culprit is that the underlying byte array in my mat is gc'ed.

If I instead create a empty mat and then fill it with data with SetSCharAt then everything is working as intended.

**Comment by brunoga** [comment]
I happen to think it is a good thing it does not do a copy. I guess an easy solution to this might be some well placed runtime.KeepAlive() One way to do this might be to set a dummy finalizer on the Mat that makes that call. Something like this (untested):

func NewMatFromBytes(rows int, cols int, mt MatType, data []byte) (Mat, error) {
	cBytes, err := toByteArray(data)
	if err != nil {
		return Mat{}, err
	}
	nm := newMat(C.Mat_NewFromBytes(C.int(rows), C.int(cols), C.int(mt), *cBytes))
        runtime.SetFinalizer(m, func(m Mat) {
                runtime.KeepAlive(data)
        })
        return nm, nil
}

**Comment by deadprogram** [comment]
The pattern in GoCV has been that if you get a `Mat` returned from any method, you have to call `Close()` on it yourself and manage the lifecycle. It is generally been a CGo antipattern to hook into the finalizer since you do not know exactly when it will be called. At least at this point seems like that is the pattern we will be sticking with. 

Separately from that, there are still some leaks that might need to be fixed internally from memory allocated by GoCV's wrapper itself.

**Comment by brunoga** [comment]
The problem here is that the lifetime of the data slice is not tied to the Mat at all. You can manage the Mat's lifecycle as much as you might want but you may still get the data slice garbage collected behind your back.

I do not think using a finalizer here is an anti-pattern. In this specific case you are just making sure that the data slice will necessarily be around at least until the Mat object is garbage collected (the finalizer will never run before it is garbage collected, of course, and here it does not really matter if it will ever run at all.

Another simpler way to have the same effect is to add a the slice itself to Mat. Something like:

type Mat struct {
        data []byte  // Nil unless the baking memory for the Mat is a byte slice.
	p C.Mat
}

Then when creating a Mat backed by a data slice, you just assign data to it *AND* add runtime.Keepalive(mat.data) to your Close() call (I am not sure this is really necessary but, conceivably, a struct member might still be garbage collected even if the struct is still around if the member is never used).

The thing here is trying to apply the law of least surprise. Without knowing an implementation detail (Mat uses the data byte slice directly from C), no one would expect Mat method calls to fail/crash because the data slice used ended up being garbage collected.

**Comment by brunoga** [comment]
FYI, I asked in the golang-nuts list and it seems that the runtime.KeepAlive(0 call would not be required as as long as the Mat instance is alive, all its attributes will also be alive. So only keeping a reference to the data slice should be enough.

**Comment by cedricve** [comment]
Is there any update on this, I'm experiencing serious memory leaks when using this function NewMatFromBytes.

**Comment by plo-** [comment]
Any news? Could you please at least leave a few notes in the documentation for NewMatFromBytes() to warn about that?

**Comment by deadprogram** [comment]
I am starting to think that `NewMatFromBytes()` should do the `copy()` and then the caller should be responsible for calling `Close()`. Otherwise we end up with somewhat unexpected behavior.

**Comment by brunoga** [comment]
> I am starting to think that `NewMatFromBytes()` should do the `copy()` and then the caller should be responsible for calling `Close()`. Otherwise we end up with somewhat unexpected behavior.

This will most likely be pretty expensive on critical paths. I still think it is simpler just to make sure the associated byte slice is not garbage collected while the Mat still exists.

Is there a specific reason why you did not like my suggestion of adding a reference to the data slice inside the Mat? Because that would solve everything at the cost of a few extra bytes per Mat.

**Comment by deadprogram** [comment]
@brunoga now that I think about it a bit further, the proposed solution of adding `data []byte` in the case of using `NewMatFromBytes()` is quite a good one. As long as `Close()` does a `m.data = nil` that should take care of the referencing as you point out.

Can you submit a PR with these changes? :)

**Comment by brunoga** [comment]
Sure. Will do it tomorrow.

**Comment by brunoga** [comment]
Pull request sent.

**Comment by deadprogram** [comment]
This was addressed by #755 and has been marked to be closed when our next release goes out. Thanks!

**Comment by deadprogram** [comment]
This was released with v0.26.0 so now closing. Thank you everyone!
